File: runtime/org.openmole.runtime/src/main/java/org/openmole/runtime/SimExplorer.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.runtime;

import org.openmole.core.batch.authentication.Authentication;
import java.io.File;
import java.util.logging.Level;

import java.util.logging.Logger;
import org.apache.commons.cli.BasicParser;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import org.eclipse.equinox.app.IApplication;
import org.eclipse.equinox.app.IApplicationContext;
import org.openmole.misc.pluginmanager.PluginManager;
import org.openmole.misc.workspace.Workspace;
import org.openmole.core.serializer.SerializerService;
import org.openmole.misc.logging.LoggerService;

public class SimExplorer implements IApplication {

    @Override
    public Object start(IApplicationContext context) throws Exception {
        try {
            String args[] = (String[]) context.getArguments().get(IApplicationContext.APPLICATION_ARGS);

            Options options = new Options();

            options.addOption("a", true, "Path to a serialized authentication to initialize.");
            options.addOption("s", true, "Base uri for the storage.");
            options.addOption("i", true, "URI of the input message.");
            options.addOption("o", true, "URI of the output message.");
            options.addOption("c", true, "Path for the communication.");
            options.addOption("p", true, "Path for plugin dir to preload.");
            options.addOption("l", true, "Local authentication mode for debug.");
            options.addOption("d", false, "Debug mode.");

            CommandLineParser parser = new BasicParser();
            CommandLine cmdLine;

            boolean debug = false;

            try {
                cmdLine = parser.parse(options, args);
            } catch (ParseException e) {
                Logger.getLogger(SimExplorer.class.getName()).severe("Error while parsing command line arguments");
                new HelpFormatter().printHelp(" ", options);
                return IApplication.EXIT_OK;
            }

            String environmentPluginDirPath = cmdLine.getOptionValue("p");
            String executionMessageURI = cmdLine.getOptionValue("i");
            String baseURI = cmdLine.getOptionValue("s");
            String communicationPath = cmdLine.getOptionValue("c");

            //System.out.println("plugin path " + environmentPluginDirPath);
            File environmentPluginDir = new File(environmentPluginDirPath);
            PluginManager.loadDir(environmentPluginDir);

            if (cmdLine.hasOption("l")) {
                Workspace.instance().password_$eq(cmdLine.getOptionValue("l"));
                debug = true;
            }

            if (cmdLine.hasOption("d")) {
                debug = true;
            }

            if (debug) {
                LoggerService.level("ALL");
            }

            //if (cmdLine.hasOption("a")) {
                /*
             * get env and init
             */

            Authentication authentication = null;
            if (cmdLine.hasOption("a")) {
                File envFile = new File(cmdLine.getOptionValue("a"));
                authentication = SerializerService.deserializeAndExtractFiles(envFile);
                authentication.initialize(false);
            }
            //if(!debug) envFile.delete();
            //}


            String outMesseageURI = cmdLine.getOptionValue("o");

            new Runtime().apply(baseURI, communicationPath, executionMessageURI, outMesseageURI, debug);

            //Be sure it is not garbage collected
            if (authentication != null) {
                authentication.getClass();
            }

        } catch (Throwable t) {
            Logger.getLogger(SimExplorer.class.getName()).log(Level.SEVERE, "Error durring runtime execution", t);
        }
        return IApplication.EXIT_OK;
    }

    @Override
    public void stop() {
    }
}

File: openmole/core/runtime/org.openmole.runtime.runtime/src/main/scala/org/openmole/runtime/runtime/SimExplorer.scala
/*
 * Copyright (C) 2012 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.runtime.runtime

import org.eclipse.equinox.app._
import scopt.immutable._
import org.openmole.misc.pluginmanager.PluginManager
import org.openmole.misc.logging.LoggerService
import org.openmole.misc.tools.io.FileUtil._
import java.io.File
import org.openmole.core.serializer.SerializerService
import org.openmole.misc.tools.service.Logger
import org.openmole.misc.workspace._
import org.openmole.core.batch.storage._
import org.openmole.core.implementation.execution.local.LocalEnvironment

class SimExplorer extends IApplication with Logger {

  override def start(context: IApplicationContext) = {
    try {
      val args = context.getArguments.get(IApplicationContext.APPLICATION_ARGS).asInstanceOf[Array[String]]

      case class Config(
        storage: Option[String] = None,
        inputMessage: Option[String] = None,
        outputMessage: Option[String] = None,
        path: Option[String] = None,
        pluginPath: Option[String] = None,
        nbThread: Option[Int] = None)

      val parser = new OptionParser[Config]("openmole", "0.x") {
        def options = Seq(
          opt("s", "storage", "Storage") {
            (v: String, c: Config) ⇒ c.copy(storage = Some(v))
          },
          opt("i", "input", "Path of the input message") {
            (v: String, c: Config) ⇒ c.copy(inputMessage = Some(v))
          },
          opt("o", "output", "Path of the output message") {
            (v: String, c: Config) ⇒ c.copy(outputMessage = Some(v))
          },
          opt("c", "path", "Path for the communication") {
            (v: String, c: Config) ⇒ c.copy(path = Some(v))
          },
          opt("p", "plugin", "Path for plugin dir to preload") {
            (v: String, c: Config) ⇒ c.copy(pluginPath = Some(v))
          },
          opt("t", "nbThread", "Number of thread for the execution") {
            (v: String, c: Config) ⇒ c.copy(nbThread = Some(v.toInt))
          })
      }

      val debug = args.contains("-d")
      val filtredArgs = args.filterNot((_: String) == "-d")
      if (debug) LoggerService.level("ALL")

      parser.parse(filtredArgs, Config()) foreach { config ⇒

        PluginManager.loadDir(new File(config.pluginPath.get))

        val storageFile = Workspace.newFile("storage", ".xml")
        val storage =
          try {
            new File(config.storage.get).copyUncompressFile(storageFile)
            SerializerService.deserializeAndExtractFiles[SimpleStorage](storageFile)
          }
          finally storageFile.delete

        LocalEnvironment.initializationNumberOfThread = config.nbThread

        new Runtime().apply(
          storage,
          config.path.get,
          config.inputMessage.get,
          config.outputMessage.get,
          debug)

      }
    }
    catch {
      case t: Throwable ⇒
        logger.log(SEVERE, "Error durring runtime execution", t)
        System.setProperty("eclipse.exitcode", "1")
    }
    IApplication.EXIT_OK

  }
  override def stop = {}

}
